### PR TITLE
Added advanced tests for listing

### PR DESCRIPTION
**Pull request by mmusil**
## JIRA:
https://issues.jboss.org/browse/RHMAP-18459
Deployed app can be found there https://testing.zeta.feedhenry.com/#projects/jyt5ylxfa6vjdz4pyg3lbrkf/apps/jyt5ylqcwe6nns2xg4fvaues/deploy

### CODE REVIEW

File: lib/db_tests.js [hunk 1]
```diff
@@ -10,6 +12,28 @@ var user2 = JSON.parse(dummyjson.parse(userTemplate));
...
 var user1Guid = null;
 
+function dbListPromisified(type, restrictions) {
+  return new Promise(function(resolve, reject){
+    var params = {
+      "act": "list",
+      "type": type
+    };
+
+    if (restrictions && typeof restrictions === "function") {
```
Comment (josemigallas) [inline lib/db_tests.js]: Is this condition really necessary? I would rather just do:
```
if (restrictions && typeof restrictions !== "function") {
    params = _.extend(params, restrictions);
}
```

File: lib/db_tests.js [hunk 2]
```diff
@@ -10,6 +12,28 @@ var user2 = JSON.parse(dummyjson.parse(userTemplate));
...
+      restrictions = null;
+    } else if (restrictions) {
+      params = _.extend(params, restrictions);
+    }
+    fh.db(params, function (error, data) {
+      if (error){
+        reject(error);
+      }
```
Comment (josemigallas) [inline lib/db_tests.js]: There are some formatting _irregularities_ here, please fix

File: lib/db_tests.js [hunk 3]
```diff
@@ -82,6 +106,8 @@ describe('Database API', function() {
   });
 
   it('should create multiple entities', function(done) {
+    user2.a = 2;
```
Comment (josemigallas) [inline lib/db_tests.js:116]: Why is this added now, were they missing? Where is `user1.a`?

File: lib/db_tests.js [hunk 4]
```diff
@@ -179,4 +205,106 @@ describe('Database API', function() {
...
+   {"lt": { "a": 2 }}       data.count === 1 ? "SUCCESS!!!!!" : "--------FAILURE--------"
+   {"le": { "a": 2 }}       data.count === 2 ? "SUCCESS!!!!!" : "--------FAILURE--------";
+   {"gt": { "a": 1 }}       data.count === 1 ? "SUCCESS!!!!!" : "--------FAILURE--------"
+   {"ge": { "a": 1 }}       data.count === 2 ? "SUCCESS!!!!!" : "--------FAILURE--------";
+   {"in": { "a": [1] }}     data.count === 1 ? "SUCCESS!!!!!" : "--------FAILURE--------"
+   {"sort": { "a": -1 }}    data.list[0].fields.a === 2 ? "SUCCESS!!!!!" : "--------FAILURE--------"
+   {"skip": 0, "limit": 1}  data.count === 1 ? "SUCCESS!!!!!" : "--------FAILURE--------"
+   */
+
+  it('advanced listing - Equals 1', function (done) {
```
Comment (josemigallas) [inline lib/db_tests.js]: Test cases should be named as "should _do whatever_", so that they explicitly describe the functionality. This one could be something like `it should filter using equal-to condition` and anyone would instantly know what is being proved.

File: lib/db_tests.js (hunk 3 above)
Comment (mmusil) [reply lib/db_tests.js:116]: those records are created in previous steps, user1 is created, modified and then deleted

File: lib/db_tests.js [hunk 5]
```diff
@@ -179,4 +205,106 @@ describe('Database API', function() {
...
+    dbListPromisified(type, {"eq": { "a": 1 }})
+      .then(function(data) {
+        expect(data.count).to.equal(1);
```
Comment (josemigallas) [inline lib/db_tests.js]: IMO the expected results should be defined as constants at the beginning of the `describe` block, instead of summarising them in a comment block. The way it's written now could easily lead to confusion and human error. What I am saying is something like:
```
const AdvancedListingRestrictions = {
  equalTo = 1;
}

[...]
expect(data.count).to.equal(AdvancedListingResults.equalTo);

File: lib/db_tests.js (hunk 3 above)
Comment (josemigallas) [reply lib/db_tests.js:116]: OK I see them now initialised on the top, my bad!

File: package.json [hunk 6]
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.3",
   "dependencies": {
     "async": "^2.5.0",
+    "bluebird": "^3.5.1",
```
Comment (psturc) [inline package.json]: Can we use native Promise?

File: lib/db_tests.js (hunk 1 above)
Comment (mmusil) [reply lib/db_tests.js]: I've been thinking about this... On one hand there could be just something like you wrote
```
if (restrictions && typeof restrictions !== "function") {
    params = _.extend(params, restrictions);
}
```
or no condition at all, because I don't see any way, how can anything else get into the variable, but on the other hand I thought, if Peter or David put this condition into their cloud app, maybe there was a reason for that 😅

File: lib/db_tests.js (hunk 5 above)
Comment (mmusil) [reply lib/db_tests.js]: this is actually good idea and I'll add it to the code 👍 do you mean something like this?
```
const AdvancedListingResults = {
    equalTo: 1,
    notEqualTo: 1,
    lessThan: 1,
    lessOrEqual: 2,
    greaterThan: 1,
    greaterOrEqual: 2,
    inArray: 1,
    sortMaxToMin: 2,
    skipFirst: 1
  };

[...]
expect(data.count).to.equal(AdvancedListingResults.equalTo);
[...]
expect(data.count).to.equal(AdvancedListingResults.notEqualTo);
[...]
```

**Review by josemigallas — APPROVED** [review]
Perfect 👌
